File: components/api/api-modules-python/src/main/resources/META-INF/dirigible/python-modules/dirigible/utils/escape.py
import html
import json
import xml.sax.saxutils as saxutils

def escape_csv(input):
    return '"' + input.replace('"', '""') + '"'

def escape_javascript(input):
    return json.dumps(input)

def escape_html3(input):
    return html.escape(input)

def escape_html4(input):
    return html.escape(input)

def escape_java(input):
    escaped = ""
    for char in input:
        if char == '\n':
            escaped += '\\n'
        elif char == '\r':
            escaped += '\\r'
        elif char == '\t':
            escaped += '\\t'
        elif char == '\\':
            escaped += '\\\\'
        elif char == '"':
            escaped += '\\"'
        elif ord(char) < 32 or ord(char) > 126:
            escaped += "\\u{:04x}".format(ord(char))
        else:
            escaped += char
    return escaped

def escape_json(input):
    return json.dumps(input)

def escape_xml(input):
    return saxutils.escape(input, entities={'"': '&quot;', "'": '&apos;'})

def unescape_csv(input):
    if input.startswith('"') and input.endswith('"'):
        input = input[1:-1]
        return input.replace('""', '"')
    return input

def unescape_javascript(input):
    return json.loads(input)

def unescape_html3(input):
    return html.unescape(input)

def unescape_html4(input):
    return html.unescape(input)

def unescape_java(input):
    unescaped = ""
    i = 0
    while i < len(input):
        char = input[i]
        if char == '\\' and i + 1 < len(input):
            next_char = input[i + 1]
            if next_char == 'n':
                unescaped += '\n'
                i += 1
            elif next_char == 'r':
                unescaped += '\r'
                i += 1
            elif next_char == 't':
                unescaped += '\t'
                i += 1
            elif next_char == '\\':
                unescaped += '\\'
                i += 1
            elif next_char == '"':
                unescaped += '"'
                i += 1
            elif next_char == "'":
                unescaped += "'"
                i += 1
            elif next_char == 'b':
                unescaped += '\b'
                i += 1
            elif next_char == 'f':
                unescaped += '\f'
                i += 1
            elif next_char == 'u' and i + 5 < len(input):
                try:
                    unicode_char = chr(int(input[i + 2:i + 6], 16))
                    unescaped += unicode_char
                    i += 5
                except ValueError:
                    unescaped += '\\' + next_char
            else:
                unescaped += '\\' + next_char
        else:
            unescaped += char
        i += 1
    return unescaped

def unescape_json(input):
    return json.loads(input)

def unescape_xml(input):
    return saxutils.unescape(input, entities={'quot': '"', 'apos': "'"})

File: components/api/api-modules-javascript/src/main/resources/META-INF/dirigible/modules/src/utils/escape.ts
/*
 * Copyright (c) 2023 SAP SE or an SAP affiliate company and Eclipse Dirigible contributors
 *
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v2.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v20.html
 *
 * SPDX-FileCopyrightText: 2023 SAP SE or an SAP affiliate company and Eclipse Dirigible contributors
 * SPDX-License-Identifier: EPL-2.0
 */

const EscapeFacade = Java.type("org.eclipse.dirigible.components.api.utils.EscapeFacade");

export class Escape{

	public static escapeCsv(input: string): string {
		return EscapeFacade.escapeCsv(input);
	};

	/**
	 * Escapes a Javascript string
	 */
	public static escapeJavascript(input: string): string {
		return EscapeFacade.escapeJavascript(input);
	};

	/**
	 * Escapes a HTML3 string
	 */
	public static escapeHtml3(input: string): string {
		return EscapeFacade.escapeHtml3(input);
	};

	/**
	 * Escapes a HTML4 string
	 */
	public static escapeHtml4(input: string): string {
		return EscapeFacade.escapeHtml4(input);
	};

	/**
	 * Escapes a Java string
	 */
	public static escapeJava(input: string): string {
		return EscapeFacade.escapeJava(input);
	};

	/**
	 * Escapes a JSON string
	 */
	public static escapeJson(input: string): string {
		return EscapeFacade.escapeJson(input);
	};

	/**
	 * Escapes a XML string
	 */
	public static escapeXml(input: string): string {
		return EscapeFacade.escapeXml(input);
	};

	/**
	 * Unescapes a CSV string
	 */
	public static unescapeCsv(input: string): string {
		return EscapeFacade.unescapeCsv(input);
	};

	/**
	 * Unescapes a Javascript string
	 */
	public static unescapeJavascript(input: string): string {
		return EscapeFacade.unescapeJavascript(input);
	};

	/**
	 * Unescapes a HTML3 string
	 */
	public static unescapeHtml3(input: string): string {
		return EscapeFacade.unescapeHtml3(input);
	};

	/**
	 * Unescapes a HTML4 string
	 */
	public static unescapeHtml4(input: string): string {
		return EscapeFacade.unescapeHtml4(input);
	};

	/**
	 * Unescapes a Java string
	 */
	public static unescapeJava(input: string): string {
		return EscapeFacade.unescapeJava(input);
	};

	/**
	 * Unescapes a JSON string
	 */
	public static unescapeJson(input: string): string {
		return EscapeFacade.unescapeJson(input);
	};

	/**
	 * Unescapes a XML string
	 */
	public static unescapeXml(input: string): string {
		return EscapeFacade.unescapeXml(input);
	};
}
